### ISSUE TITLE
Context API: place external data object

### ISSUE THREAD

**Issue by eirrgang** [kassonlab/gmxapi#96]
Since we are getting away from explicit reference to the filesystem, especially in the WorkSpec, we need a way for helper functions and chains of Contexts to place data objects like TPR input files in the right place once the unique tag for the WorkSpec is known.

**Comment by eirrgang** [comment]
In some situations, a conceptually unified data set may actually be split across multiple files. Ref. #86 

To manage simulation microstates, we currently need both TPR and CPT files. And, of course, the TPR file contains more than just micro state information, so conceptually discrete data sets may be included in the _same_ file.

This is definitely going to get interesting... It may be preferable to get away from internal use of the common GROMACS files and use them only as a data interchange mechanism. We want an abstraction from the filesystem anyway, so I guess it is time to think about what the filesystem-backed data objects in different context implementations will look like. Using TNG for file backed data on Unix filesystems may be optimal for the libgromacs library Context.

**Comment by peterkasson** [comment]
As noted elsewhere, TPR and CPT correspond to two different abstracted entities:
1.  run input
2.  simulation state

**Comment by eirrgang** [comment]
Yes. Neither one corresponds completely to an abstraction of complete microstate in all cases, so this issue exists to track the concept of starting a simulation from an abstraction of a microstate, regardless of whether the simulation trajectory has hysteresis yet.

If we do use the conceptual abstraction of a microstate, it is important to be clear about what it is an aggregate of, but it may be preferable not to lump the constituents at all. I don't have a strong opinion either way, but I'm leaning towards the sense that hidden degrees of freedom, integrator/thermostat/barostat state, constraint state (whether initialized or not) is all part of the microstate in that it affects the Markov chain, but that many users will find it preferable not to have to think about the implementation details and instead just bundle up the "microstate." However, that means the MD method affects the definition of a microstate and leaves open the question of how the API should handle passing microstate data between systems with different degrees of freedom / ensembles, or whatever (such as minimizing in NVE and switching to. NVT).

To avoid glitches in trajectories due to interrupted internal state, I expect the short term management of points in a trajectory to require intelligent handling of both TPR and checkpoint files. I need more discussion / feedback to get a sense of the appropriate long-term abstraction(s) and how best to implement them.

This task will be resolved by guaranteeing the presence of input data for an operation. For a simulation operation whose input is the output of another simulation, this means assuring a copy of both the TPR and checkpoint files in the working directory. Cleverer implementation (minimizing redundant on-disk data) requires workspec schema 0.2, specifically #65 , #76.

To continue the discussion of data graph edge types (how specific, how specified, etc), we could update #65 or start a new issue. #34 and #36 probably also deserve an update.
